### PR TITLE
[sail] Add new port

### PR DESCRIPTION
**Pull request by HappySeaFox**
**Describe the pull request**

Draft PR to test a new port: SAIL.

- What does your PR fix? Fixes #
N/A.

- Which triplets are supported/not supported? Have you updated the CI baseline?
Only Windows non-static platform is supported. CI baseline is not updated.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes.

### CODE REVIEW

File: ports/sail/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  "port-version": 1,
+  "description": "The missing small and fast image decoding library for humans (not for machines)",
+  "homepage": "https://github.com/smoked-herring/sail",
+  "dependencies": [
+    "giflib",
+    "libjpeg-turbo",
+    "libpng",
+    "tiff"
+  ]
+}
```
Comment (NancyLi1013) [inline ports/sail/vcpkg.json:13]: `libjpeg-turbo` is one of dependencies for `tiff`. So `tiff `might be enough here. Could you please help confirm this?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/sail/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,39 @@
+vcpkg_fail_port_install(ON_TARGET "UWP" "OSX" "Linux")
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO smoked-herring/sail
+    REF 3672fd1
+    SHA512 a02dfcd8eeedd6400c6041542579149a95d9933ec70e2fdfad71de847af3eb3f268b37529be14faefe6c71bf96ed11f6405399c64eac7a154582d86195fa361d
+    HEAD_REF feature/vcpkg
```
[the remark this answers is not in the document]
Comment (smoked-herring) [reply ports/sail/portfile.cmake:8]: `feature/vcpkg` is an experiment to add vcpkg support. When `feature/vcpkg` is ready, I'll merge it into the master, and update the port file as well as the PR.

File: ports/sail/vcpkg.json (hunk 1 above)
Comment (smoked-herring) [reply ports/sail/vcpkg.json:12]: I directly depend on `libjpeg-turbo` as well. That's why it's here.

File: ports/sail/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,39 @@
+vcpkg_fail_port_install(ON_TARGET "UWP" "OSX" "Linux")
```
[the remark this answers is not in the document]
Comment (smoked-herring) [reply ports/sail/portfile.cmake:1]: You're absolutely right, SAIL supports OSX and Linux, but natively. I.e. w/o using vcpkg. I plan to extend vcpkg support for Linux/OSX later. For now, there are some compilation issues and vpkg peculiarities (due to static linking on Linux) that stop me from doing that.

File: scripts/ci.baseline.txt [hunk 4]
```diff
@@ -1346,6 +1346,13 @@ ryu:x64-uwp=fail
 ryu:x64-windows-static=fail
 ryu:x86-windows=fail
 ryu::arm64-windows=fail
+# SAIL doesn't support static compilation yet
+sail:x64-windows-static=skip
```
Comment (NancyLi1013) [inline scripts/ci.baseline.txt:1350]: Does the port not support static build? If so, you can add this in `Supports: !static`. For uwp, you can handle it in the same way.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/sail/portfile.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (smoked-herring) [reply ports/sail/portfile.cmake:1]: I think not in this PR. In this PR I target Windows only.

File: ports/sail/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+# Remove duplicate files
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include
+                    ${CURRENT_PACKAGES_DIR}/debug/share)
+
+# Move codecs
+file(RENAME ${CURRENT_PACKAGES_DIR}/lib/sail ${CURRENT_PACKAGES_DIR}/bin/sail)
+file(REMOVE ${CURRENT_PACKAGES_DIR}/lib)
+
+file(RENAME ${CURRENT_PACKAGES_DIR}/debug/lib/sail ${CURRENT_PACKAGES_DIR}/debug/bin/sail)
+file(REMOVE ${CURRENT_PACKAGES_DIR}/debug/lib)
```
[the remark this answers is not in the document]
Comment (smoked-herring) [reply ports/sail/portfile.cmake:33]: Oops, it looks like a mistake!

File: scripts/ci.baseline.txt (hunk 4 above)
Comment (smoked-herring) [reply scripts/ci.baseline.txt]: Thanks! I'll commit a fix.

**Comment by smoked-herring** [comment]
Updated REF to master. I think it's ready for a review.

**Comment by smoked-herring** [comment]
@NancyLi1013 Could you please review once more?

**Comment by NancyLi1013** [comment]
@smoked-herring 

Sorry for the delay.

I noticed that there are also some *.pc files in` lib/pkgconfig`. So it would be better to add` vcpkg_fixup_pkgconfig()`.

**Comment by smoked-herring** [comment]
@NancyLi1013 

Thanks for the review!

>I noticed that there were also some *.pc files in lib/pkgconfig. So it would be better to add vcpkg_fixup_pkgconfig().

Added `vcpkg_fixup_pkgconfig()`.

>I also found that there were some *.info and *.dll files listed in bin/sail/codes. Do you know what these files are used for?

These are dynamically loaded codecs (plugins). They provide actual image decoding capabilities. Static build is not supported for now, so codecs are distributed as standalone DLL plugins.

>For the usage of sail, have you tested it?

Yes, there are demo apps. I tested all of them against the port, everything works fine. The only caveat is that a user needs to copy codecs (`bin/sail/codecs`) and their dependencies by hand into the application directory. I also added a `usage` file to inform the user about this peculiarity.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Pull request merged**